tests/dialects/test_stablehlo.py

```diff
@@ -0,0 +1,38 @@
+import platform
+from typing import TYPE_CHECKING
+
+from xdsl.dialects.builtin import AnyTensorType, IntegerType, TensorType
+from xdsl.dialects.stablehlo import (
+    AddOp,
+    _abstract_operation_class_factory,  # pyright: ignore[reportPrivateUsage]
+)
+from xdsl.irdl import VarConstraint, base
+from xdsl.utils.test_value import TestSSAValue
+
+has_reveal_type = TYPE_CHECKING and int(platform.python_version_tuple()[1]) >= 11
+if has_reveal_type:
+    # reveal_type is only supported on python 3.11 and above
+    # https://docs.python.org/3/library/typing.html#typing.reveal_type
+    from typing import reveal_type
```

Comment:
Just fiddled with this locally, I think something like this is actually what we want here:

```python
def test_bla():
    from typing_extensions import assert_type

    assert_type(
        AddiOp.result,
        OpResult[
            IndexType
            | IntegerType
            | VectorType[IndexType | IntegerType]
            | TensorType[IndexType | IntegerType]
        ],
    )
```

Thank you for iterating on this!

Reply:
Unlike reveal_type, which always fails at runtime, assert_type asserts only when the type mismatches

Reply:
Ok, will try tomorrow. Haha. Off to dinner. Thanks for the tips!

Reply:
Ok, this shows that the type does not propagate as you expect with pyright. I only got OpResult[Attribute] to work, but not OpResult[TensorType[Attribute]]. I did try the code in main and it showed OpResult[TensorType[Attribute]]. Feel free to close or suggest something else.

Reply:
Yeah in this case I think something like generating the code from the tablegen definition would be better than a factory like this, let's close it for now.